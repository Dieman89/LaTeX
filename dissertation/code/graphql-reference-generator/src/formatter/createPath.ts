import {
  isInterfaceType,
  isObjectType,
  isEnumType,
  isScalarType,
  isUnionType,
  isInputType,
  isListType,
  isNonNullType,
  GraphQLType,
  DirectiveNode,
} from "graphql";
import {
  DIRECTIVES,
  ENUMS,
  INPUTS,
  INTERFACES,
  OBJECTS,
  SCALARS,
  UNIONS,
} from "../constants";
import { toKebabCase } from "../utils/toKebabCase";

export const createPathToItem = (type: GraphQLType | DirectiveNode): string => {
  if (isListType(type) || isNonNullType(type)) {
    return createPathToItem(type.ofType);
  }
  if (isScalarType(type)) {
    return `/${SCALARS}/${toKebabCase(type.name)}`;
  }
  if (isObjectType(type)) {
    return `/${OBJECTS}/${toKebabCase(type.name)}`;
  }
  if (isEnumType(type)) {
    return `/${ENUMS}/${toKebabCase(type.name)}`;
  }
  if (isInterfaceType(type)) {
    return `/${INTERFACES}/${toKebabCase(type.name)}`;
  }
  if (isUnionType(type)) {
    return `/${UNIONS}/${toKebabCase(type.name)}`;
  }
  if (isInputType(type)) {
    return `/${INPUTS}/${toKebabCase(type.name)}`;
  }
  return `/${DIRECTIVES}/${toKebabCase(type.name.value)}`;
};
