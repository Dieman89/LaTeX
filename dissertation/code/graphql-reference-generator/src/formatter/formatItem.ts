import { GraphQLDirectiveItem, GraphQLItem, GraphQLRawItem } from "../types";
import {
  GraphQLArgument,
  GraphQLField,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLInputObjectType,
  print,
  ASTNode,
  DirectiveNode,
  GraphQLDirective,
} from "graphql";
import { createPathToItem } from "./createPath";

export const formatItem = (graphqlType: GraphQLRawItem): GraphQLItem => {
  return {
    name: graphqlType.name,
    description: graphqlType.description || "",
    type: hasType(graphqlType) ? graphqlType.type.toString() : null,
    printed: printASTNode(graphqlType.astNode as ASTNode),
    arguments: hasArgs(graphqlType)
      ? Object.values(graphqlType.args).map(formatItem)
      : [],
    fields: hasFields(graphqlType)
      ? Object.values(graphqlType.getFields()).map(formatItem)
      : [],
    path: hasType(graphqlType)
      ? createPathToItem(graphqlType.type).toLowerCase()
      : null,
    directives: hasDirectives(graphqlType)
      ? createDirectives(graphqlType.astNode.directives)
      : [],
  };
};

export function printASTNode(astNode: ASTNode) {
  const pattern = new RegExp(/"""([\s\S]*?)"""\n/, "g");
  return astNode
    ? print(astNode).replace(pattern, "").trim().replace(/^ +/gm, "")
    : null;
}

export function createDirectives(
  nodes: Array<DirectiveNode>
): Array<GraphQLDirectiveItem> {
  return nodes.map((item: DirectiveNode) => {
    return {
      name: item.name.value,
      path: createPathToItem(item),
    };
  });
}

function hasType(
  type: any
): type is GraphQLArgument | GraphQLField<any, any, any> {
  return !!type.type;
}

function hasArgs(
  type: any
): type is GraphQLField<any, any, any> | GraphQLDirective {
  return !!type.args;
}

function hasFields(
  type: any
): type is GraphQLObjectType | GraphQLInterfaceType | GraphQLInputObjectType {
  return !!type.getFields;
}

function hasDirectives(
  type: any
): type is { astNode: { directives: DirectiveNode[] } } {
  return !!(type.astNode?.directives?.length > 0);
}
