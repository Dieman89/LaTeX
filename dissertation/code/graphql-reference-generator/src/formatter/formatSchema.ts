import {
  GraphQLDirective,
  GraphQLNamedType,
  GraphQLObjectType,
  GraphQLSchema,
  isEnumType,
  isInputType,
  isInterfaceType,
  isObjectType,
  isScalarType,
  isUnionType,
} from "graphql";
import {
  ENUMS,
  INPUTS,
  INTERFACES,
  MUTATIONS,
  OBJECTS,
  QUERIES,
  SCALARS,
  SUBSCRIPTIONS,
  UNIONS,
  DIRECTIVES,
} from "../constants";
import { formatItem } from "./formatItem";
import { GraphQLReferenceGroup } from "../types";

export function formatSchema(schema: GraphQLSchema) {
  return {
    ...Object.values(schema.getTypeMap())
      .filter((type) => !type.name.startsWith("__"))
      .reduce(typeReducer, initialItemData),
    directives: schema.getDirectives().map(formatItem),
  };
}

const initialItemData: GraphQLReferenceGroup = {
  [QUERIES]: [],
  [MUTATIONS]: [],
  [SUBSCRIPTIONS]: [],
  [OBJECTS]: [],
  [ENUMS]: [],
  [SCALARS]: [],
  [INTERFACES]: [],
  [UNIONS]: [],
  [INPUTS]: [],
  [DIRECTIVES]: [],
};

function typeReducer(
  acc: GraphQLReferenceGroup,
  type: GraphQLNamedType | GraphQLDirective
): GraphQLReferenceGroup {
  if (isObjectType(type)) {
    switch (type.name) {
      case "Query":
        return {
          ...acc,
          [QUERIES]: createRootObjectArray(type),
        };

      case "Mutation":
        return {
          ...acc,
          [MUTATIONS]: createRootObjectArray(type),
        };

      case "Subscription":
        return {
          ...acc,
          [SUBSCRIPTIONS]: createRootObjectArray(type),
        };

      default:
        return {
          ...acc,
          [OBJECTS]: [...acc[OBJECTS], formatItem(type)],
        };
    }
  }

  if (isEnumType(type)) {
    return {
      ...acc,
      [ENUMS]: [...acc[ENUMS], formatItem(type)],
    };
  }

  if (isScalarType(type)) {
    return {
      ...acc,
      [SCALARS]: [...acc[SCALARS], formatItem(type)],
    };
  }

  if (isInterfaceType(type)) {
    return {
      ...acc,
      [INTERFACES]: [...acc[INTERFACES], formatItem(type)],
    };
  }

  if (isUnionType(type)) {
    return {
      ...acc,
      [UNIONS]: [...acc[UNIONS], formatItem(type)],
    };
  }

  if (isInputType(type)) {
    return {
      ...acc,
      [INPUTS]: [...acc[INPUTS], formatItem(type)],
    };
  }

  return acc;
}

function createRootObjectArray(type: GraphQLObjectType) {
  return Object.values(type.getFields())
    .map(formatItem)
    .map((formattedType) => ({
      ...formattedType,
      isRootField: true,
    }));
}
