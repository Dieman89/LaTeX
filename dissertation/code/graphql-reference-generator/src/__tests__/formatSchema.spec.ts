import { buildSchema } from "graphql";
import { formatSchema } from "../formatter/formatSchema";

describe("formatSchema", () => {
  it("takes a GraphQLSchema and produces a correctly formatted object", () => {
    const schema = buildSchema(`
      schema {
        query: Query
        mutation: Mutation
        subscription: Subscription
      }

      """
      This is a query type
      """
      type Query {
        """
        Test query field description
        """
        testQueryField(testArgument: ID!): TestObject
      }

      """
      Test mutation description
      """
      type Mutation {
        """
        Test mutation field description
        """
        testMutationField(testInput: TestInput): TestObject
      }

      """
      Test subscription description
      """
      type Subscription {
        """
        Test subscription field description
        """
        testSubscription(testArgument: String): TestObject
      }

      """
      Test input description
      """
      input TestInput {
        """
        Test input field description
        """
        id: String
      }

      """
      Test object type description
      """
      type TestObject {
        """
        Test id field description
        """
        testObjectField: String!
      }

      enum TestEnum {
        TEST1
        TEST2
      }

      scalar TestScalar

      interface TestInterface {
        id: String
      }

      union TestUnion = String | Int
    `);

    const output = formatSchema(schema);

    expect(output).toMatchSnapshot();
  });
});
