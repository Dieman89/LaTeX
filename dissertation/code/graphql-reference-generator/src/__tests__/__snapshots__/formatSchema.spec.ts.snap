// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatSchema takes a GraphQLSchema and produces a correctly formatted object 1`] = `
Object {
  "directives": Array [
    Object {
      "arguments": Array [
        Object {
          "arguments": Array [],
          "description": "Included when true.",
          "directives": Array [],
          "fields": Array [],
          "name": "if",
          "path": "/scalars/boolean",
          "printed": null,
          "type": "Boolean!",
        },
      ],
      "description": "Directs the executor to include this field or fragment only when the \`if\` argument is true.",
      "directives": Array [],
      "fields": Array [],
      "name": "include",
      "path": null,
      "printed": null,
      "type": null,
    },
    Object {
      "arguments": Array [
        Object {
          "arguments": Array [],
          "description": "Skipped when true.",
          "directives": Array [],
          "fields": Array [],
          "name": "if",
          "path": "/scalars/boolean",
          "printed": null,
          "type": "Boolean!",
        },
      ],
      "description": "Directs the executor to skip this field or fragment when the \`if\` argument is true.",
      "directives": Array [],
      "fields": Array [],
      "name": "skip",
      "path": null,
      "printed": null,
      "type": null,
    },
    Object {
      "arguments": Array [
        Object {
          "arguments": Array [],
          "description": "Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted using the Markdown syntax, as specified by [CommonMark](https://commonmark.org/).",
          "directives": Array [],
          "fields": Array [],
          "name": "reason",
          "path": "/scalars/string",
          "printed": null,
          "type": "String",
        },
      ],
      "description": "Marks an element of a GraphQL schema as no longer supported.",
      "directives": Array [],
      "fields": Array [],
      "name": "deprecated",
      "path": null,
      "printed": null,
      "type": null,
    },
    Object {
      "arguments": Array [
        Object {
          "arguments": Array [],
          "description": "The URL that specifies the behaviour of this scalar.",
          "directives": Array [],
          "fields": Array [],
          "name": "url",
          "path": "/scalars/string",
          "printed": null,
          "type": "String!",
        },
      ],
      "description": "Exposes a URL that specifies the behaviour of this scalar.",
      "directives": Array [],
      "fields": Array [],
      "name": "specifiedBy",
      "path": null,
      "printed": null,
      "type": null,
    },
  ],
  "enums": Array [
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [],
      "name": "TestEnum",
      "path": null,
      "printed": "enum TestEnum {
TEST1
TEST2
}",
      "type": null,
    },
  ],
  "inputs": Array [
    Object {
      "arguments": Array [],
      "description": "Test input description",
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": "Test input field description",
          "directives": Array [],
          "fields": Array [],
          "name": "id",
          "path": "/scalars/string",
          "printed": "id: String",
          "type": "String",
        },
      ],
      "name": "TestInput",
      "path": null,
      "printed": "input TestInput {
id: String
}",
      "type": null,
    },
  ],
  "interfaces": Array [
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": "",
          "directives": Array [],
          "fields": Array [],
          "name": "id",
          "path": "/scalars/string",
          "printed": "id: String",
          "type": "String",
        },
      ],
      "name": "TestInterface",
      "path": null,
      "printed": "interface TestInterface {
id: String
}",
      "type": null,
    },
  ],
  "mutations": Array [
    Object {
      "arguments": Array [
        Object {
          "arguments": Array [],
          "description": "",
          "directives": Array [],
          "fields": Array [],
          "name": "testInput",
          "path": "/inputs/test-input",
          "printed": "testInput: TestInput",
          "type": "TestInput",
        },
      ],
      "description": "Test mutation field description",
      "directives": Array [],
      "fields": Array [],
      "isRootField": true,
      "name": "testMutationField",
      "path": "/objects/test-object",
      "printed": "testMutationField(testInput: TestInput): TestObject",
      "type": "TestObject",
    },
  ],
  "objects": Array [
    Object {
      "arguments": Array [],
      "description": "Test object type description",
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": "Test id field description",
          "directives": Array [],
          "fields": Array [],
          "name": "testObjectField",
          "path": "/scalars/string",
          "printed": "testObjectField: String!",
          "type": "String!",
        },
      ],
      "name": "TestObject",
      "path": null,
      "printed": "type TestObject {
testObjectField: String!
}",
      "type": null,
    },
  ],
  "queries": Array [
    Object {
      "arguments": Array [
        Object {
          "arguments": Array [],
          "description": "",
          "directives": Array [],
          "fields": Array [],
          "name": "testArgument",
          "path": "/scalars/id",
          "printed": "testArgument: ID!",
          "type": "ID!",
        },
      ],
      "description": "Test query field description",
      "directives": Array [],
      "fields": Array [],
      "isRootField": true,
      "name": "testQueryField",
      "path": "/objects/test-object",
      "printed": "testQueryField(testArgument: ID!): TestObject",
      "type": "TestObject",
    },
  ],
  "scalars": Array [
    Object {
      "arguments": Array [],
      "description": "The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`\\"4\\"\`) or integer (such as \`4\`) input value will be accepted as an ID.",
      "directives": Array [],
      "fields": Array [],
      "name": "ID",
      "path": null,
      "printed": null,
      "type": null,
    },
    Object {
      "arguments": Array [],
      "description": "The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
      "directives": Array [],
      "fields": Array [],
      "name": "String",
      "path": null,
      "printed": null,
      "type": null,
    },
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [],
      "name": "TestScalar",
      "path": null,
      "printed": "scalar TestScalar",
      "type": null,
    },
    Object {
      "arguments": Array [],
      "description": "The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1.",
      "directives": Array [],
      "fields": Array [],
      "name": "Int",
      "path": null,
      "printed": null,
      "type": null,
    },
    Object {
      "arguments": Array [],
      "description": "The \`Boolean\` scalar type represents \`true\` or \`false\`.",
      "directives": Array [],
      "fields": Array [],
      "name": "Boolean",
      "path": null,
      "printed": null,
      "type": null,
    },
  ],
  "subscriptions": Array [
    Object {
      "arguments": Array [
        Object {
          "arguments": Array [],
          "description": "",
          "directives": Array [],
          "fields": Array [],
          "name": "testArgument",
          "path": "/scalars/string",
          "printed": "testArgument: String",
          "type": "String",
        },
      ],
      "description": "Test subscription field description",
      "directives": Array [],
      "fields": Array [],
      "isRootField": true,
      "name": "testSubscription",
      "path": "/objects/test-object",
      "printed": "testSubscription(testArgument: String): TestObject",
      "type": "TestObject",
    },
  ],
  "unions": Array [
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [],
      "name": "TestUnion",
      "path": null,
      "printed": "union TestUnion = String | Int",
      "type": null,
    },
  ],
}
`;
